brave/browser/net/proxy/proxy_config_service_tor.cc

```
@@ -0,0 +1,162 @@
+                      tor_proxy.begin() + url.scheme.begin + url.scheme.len);
+      }
+      if (url.host.is_valid()) {
+        host_ =
+          std::string(tor_proxy.begin() + url.host.begin,
+                      tor_proxy.begin() + url.host.begin + url.host.len);
+      }
+      if (url.port.is_valid()) {
+        port_ =
+          std::string(tor_proxy.begin() + url.port.begin,
+                      tor_proxy.begin() + url.port.begin + url.port.len);
+      }
+
+      content::ServiceManagerConnection::GetForProcess()
+        ->GetConnector()
+        ->BindInterface(tor::mojom::kTorServiceName,
+                        &tor_launcher_);
+
+      tor_launcher_.set_connection_error_handler(
+        base::BindOnce(&ProxyConfigServiceTor::OnTorLauncherCrashed,
+                       base::Unretained(this)));
+
+      BrowserThread::PostTask(
+          BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+          base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                     base::Unretained(this)));
+    }
+    config_.proxy_rules().ParseFromString(std::string(scheme_ + "://" + host_
+      + ":" + port_));
+}
+
+void ProxyConfigServiceTor::OnTorCrashed(int32_t pid) {
+  LOG(ERROR) << "Tor Process(" << pid << ") Crashed";
+  BrowserThread::PostTask(
+      BrowserThread::PROCESS_LAUNCHER, FROM_HERE,
+      base::Bind(&ProxyConfigServiceTor::LaunchTorProcess,
+                 base::Unretained(this)));
+}
+
+ProxyConfigServiceTor::~ProxyConfigServiceTor() {}
+
+void ProxyConfigServiceTor::LaunchTorProcess() {
+  std::vector<std::string> args;
+  args.push_back("--ignore-missing-torrc");
+  args.push_back("-f");
+  args.push_back("/nonexistent");
+  args.push_back("--defaults-torrc");
+  args.push_back("/nonexistent");
+  args.push_back("--SocksPort");
+  args.push_back(host_ + ":" + port_);
+  base::FilePath user_data_dir;
+  PathService::Get(chrome::DIR_USER_DATA, &user_data_dir);
+  if (!user_data_dir.empty()) {
```

Can we fail noisily in the case that the user_data_dir is unset?  It seems imprudent to let tor pick somewhere to scribble outside ~/.config/brave if it doesn't exist.  Not a big issue; if this is not convenient then don't worry about it.

---

It is unlikely to happen and even if it does bunch of browser functions will not work

---

the browser won't launch at all if user_data_dir is not set